### PR TITLE
Headerification and root juggling

### LINKED ISSUE #380: Parent root reconstruction + granular state roots + body/header segregation
### What is wrong?
Part of Trinity #379
Spec PRs:
   - Merged PR: ethereum/eth2.0-specs#711
   - Origin PR: ethereum/eth2.0-specs#649

### PR DESCRIPTION
**Pull request by ralexstokes**
### What was wrong?

Closes #380.

### How was it fixed?

Implementing the changes described in that issue.

edit: sorry this has become a monster PR, happy to split it up if that is helpful for reviewing

changes inside:

- [x]  add `BeaconBlockHeader` type
- [x]  add `get_temporary_block_header`
- [x]  add `get_state_root`
- [x]  add `latest_state_roots`
- [x]  update the `SignatureDomain` enum
- [x]  proposer slashing has two conflicting block headers now instead of proposal
- [x]  move `randao_reveal` and `eth1_data` into block body
- [x]  combine batched roots into `historical_roots` in the `BeaconState`
- [x]  update name for `genesis_eth1_data`
- [x]  update `process_slot_transition` to mirror the spec function `advance_slot`
- [x]  add function `process_block_header`
- [x]  update `process_proposer_slashing`:
- [x]   - moved ahead of spec#711 to ensure that header's epochs are the same (rather than slot)
- [x]   - assert roots of headers are different
- [x]   - updates to verify headers (rather than `Proposals`)
- [x]  various book-keeping for the above changes per-{slot, block, epoch}
- [x] add `latest_block_header` to `BeaconState`

and 
- [ ] updates to tests or new tests as required

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

File: eth2/beacon/enums.py [hunk 1]
```diff
@@ -2,8 +2,9 @@
...
-    DOMAIN_ATTESTATION = 1
-    DOMAIN_PROPOSAL = 2
-    DOMAIN_EXIT = 3
-    DOMAIN_RANDAO = 4
+    DOMAIN_BEACON_BLOCK = 0
+    DOMAIN_RANDAO = 1
+    DOMAIN_ATTESTATION = 2
+    DOMAIN_DEPOSIT = 3
+    DOMAIN_VOLUNTARY_EXIT = 4
+    DOMAIN_TRANSFER = 4
```
Comment (hwwhww) [inline eth2/beacon/enums.py]:
```suggestion
    DOMAIN_TRANSFER = 5
```

File: eth2/beacon/state_machines/forks/serenity/epoch_processing.py [hunk 2]
```diff
@@ -1290,6 +1291,21 @@ def eligible(index: ValidatorIndex) -> bool:
...
 
+def _update_historical_roots(state: BeaconState,
+                             next_epoch: Epoch,
+                             config: BeaconConfig) -> BeaconState:
+    updated_historical_roots = state.historical_roots
+    epochs_per_historical_root = config.SLOTS_PER_HISTORICAL_ROOT // config.SLOTS_PER_EPOCH
+    should_update_historical_roots = next_epoch % epochs_per_historical_root == 0
+    if should_update_historical_roots:
+        roots = state.latest_block_roots + state.latest_state_roots
+        updated_historical_roots += (get_merkle_root(roots),)
```
Comment (hwwhww) [inline eth2/beacon/state_machines/forks/serenity/epoch_processing.py]: Note: https://github.com/ethereum/trinity/issues/419 will replace `get_merkle_root` with SSZ `hash_tree_root`

File: eth2/beacon/enums.py (hunk 1 above)
Comment (ralexstokes) [reply eth2/beacon/enums.py]: oh wow good catch ^_^

File: eth2/beacon/state_machines/forks/serenity/epoch_processing.py (hunk 2 above)
Comment (ralexstokes) [reply eth2/beacon/state_machines/forks/serenity/epoch_processing.py]: cool -- i suppose i'm waiting a bit to see the outcome of our serialization conversations but will keep this in mind, esp if i end up doing the migration to make sure we are using all the right roots in all the right places :)

**Comment by ralexstokes** [comment]
ok -- i have updated the description above but basically this became a mega-PR :(

happy to split it up if that is helpful :)

i have made a final pass and once i get the existing tests passing and add new ones for some of the new functionality i haven't tested yet, then this will be ready for review/merge!! as far as I can tell we will have implemented almost everything (skipping `Deposits` and `Transfers` for now it seems) in https://github.com/ethereum/eth2.0-specs/pull/711 via all the PRs listed in #379.

**Comment by ralexstokes** [comment]
looks like this is blocked on a proper `hash_tree_root` so I'll see if I can start to address enough of #419 to get the tests passing

**Comment by ralexstokes** [comment]
Looks like there are some bug fixes on top of spec PR 711.

In particular, will need to incorporate this before we can get this merged:

https://github.com/ethereum/eth2.0-specs/pull/816

**Comment by ralexstokes** [comment]
Breaking this PR up into #455, #456, #457, #458 and one more PR for the addition of the block header data type. Will close this PR once it has been superseded by the others.

**Comment by hwwhww** [comment]
@ralexstokes 
ethereum/eth2.0-specs#816 is merged and I opened a Trinity corresponding issue here: #466

**Comment by ralexstokes** [comment]
Closing in favor of #470 and the other PRs here: https://github.com/ethereum/trinity/pull/397#issuecomment-476411274

**Pull request closed without merging**